SYNTHESIZER: Add quick filter into goto-synthesizer

qinheping opened the pull request:
<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.

If your PR fixes a bug, include the regression test(s) in the same commit as the bug fix. Else, keep commits small and orthogonal, possibly placing tests in commits of their own.
--->
The quick filter check if a candidate invariant (predicates of type `exprt`) is consistent with the previous seen examples by evaluating the candidate on the examples (valuations of variables collected from CBMC trace). The idea is that the quick filter is more efficient than CBMC for checking if a candidate is incorrect.

So the synthesizer can use the quick filter to rule out incorrect candidates in an early stage before sending them to CBMC.

In detail, we say a candidate is consistent with a positive (negative) example if the evaluation of the candidate on it is `true` (`false`). If a candidate is inconsistent with any example, it cannot be a valid loop invariant predicate.

Examples are collected in the verification stage. In a CBMC trace, the valuations of variables upon the entry of the loop (loop_entry) must satisfy the loop invariants, so that they are positive examples. Oppositely, valuations of variables in the havoced iteration must not satisfy the loop invariants, so that they are negative examples.

The quick filter significantly reduce the number of candidate sent to CBMC. As an example, in the benchmark `loop_contracts_synthesis_04`, 60 out of 67 bad candidates
are ruled out be the quick filter.

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [x] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [x] My commit message includes data points confirming performance improvements (if claimed).
- [ ] My PR is restricted to a single feature or bugfix.
- [ ] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

qinheping commented:
I also modify the benchmark `loop_contracts_synthesis_03` to avoid long running time due to the bug reported in https://github.com/diffblue/cbmc/issues/7978.

tautschnig commented on src/goto-synthesizer/cegis_evaluator.cpp:
```diff
@@ -0,0 +1,343 @@
+      }
+    }
+
+    // Typecast `result` to the type of `expr`.
+    auto tmp_expr = from_integer(result, expr.type());
+    to_integer(tmp_expr, result);
+    return result;
```
This piece of code (repeated below) looks strange. Are you really just doing a modulo operation here?

qinheping replied:
The `result` is the evaluation of `expr`. `result` may be larger than the max of the type of `expr` and hence overflow. So I use `from_integer` and `to_integer` here to typecast the result to the type of `expr` to avoid overflow.

tautschnig commented on src/goto-synthesizer/cegis_evaluator.h:
```diff
@@ -0,0 +1,56 @@
+  cegis_evaluatort(
+    const exprt &expr,
+    const std::vector<cext> &cexs,
+    const messaget &log)
+    : checked_expr(expr), cexs(cexs), log(log)
```
I don't think you actually use `log`, do you?

tautschnig commented on src/goto-synthesizer/enumerative_loop_contracts_synthesizer.cpp:
```diff
@@ -26,9 +26,11 @@ Author: Qinheping Hu
+#include "cegis_evaluator.h"
 #include "expr_enumerator.h"
 
+#include <iostream>
```
Debugging left-over?

qinheping committed:
Fix contains_symbol_prefix in goto-synthesizer

The function should scan all sub-expressions `it` of the given expression `expr` and check if `*it` has a given prefix.

qinheping committed:
Add quick filter into goto-synthesizer

The quick filter check if a candidate is consistent with the previous seen examples by evaluating the candidate on the examples.
The idea is that quick filter is more efficient than CBMC for checking if a candidate is incorrect.
So the synthesizer can use the quick filter to rule out incorrect candidates in an early stage before sending them to CBMC.

In detail, we say a candidate is consistent with a positive (negative) example if the evaluation of the candidate on it is true (false).
If a candidate is inconsistent with any example, it cannot be a valid loop invariant predicate.

Examples are collected in the verification stage. In a CBMC trace, the valuations of variables upon the entry of the loop (loop_entry)
must satisfy the loop invariants, so that they are positive examples. Oppositely, valuations of variables in the havoced iteration must not satisfy
the loop invariants, so that they are negative examples.

The quick filter sinificantly reduce the number of candidate sent to CBMC. As an example, in the benchmark `loop_contracts_synthesis_04`, 61 out of 67 bad candidates
are ruled out be the quick filter.
